aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
   return result;
 }
 
+static Tensor& linalg_vector_norm_impl(Tensor& result, const Tensor& self, optional<Scalar> opt_ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  auto ord = opt_ord.value_or(2).toDouble();
+
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "norm only supports CPU AND CUDA device type, got: ", self.device().type());
+  TORCH_CHECK(self.layout() == Layout::Strided,
+              "norm only supports strided layout, got: ", self.layout());
+
+  ScalarType in_dtype = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  TORCH_CHECK(
+      at::isFloatingType(in_dtype) || at::isComplexType(in_dtype),
+      "Can only calculate the norm of floating point and complex dtypes. Got ",
+      toString(in_dtype),
+      " instead.");
+
+  IntArrayRef dim = opt_dim.value_or(IntArrayRef{});
+
+  // For NumPy compatibility, we can only perform order +/-infinity reduction
+  // (max/min) on a tensor with zero elements if the dimensions to reduce are
+  // nonzero. Otherwise, throw an error.
+  if (self.numel() == 0 && std::abs(ord) == INFINITY) {
+    bool has_identity = true;
+
+    if (dim.size() == 0) {
+      has_identity = false;
+    } else {
+      for (int64_t dim_num : dim) {
+        if (self.size(dim_num) == 0) {
+          has_identity = false;
+          break;
+        }
+      }
+    }
+    if (!has_identity) {
+      TORCH_CHECK(ord != INFINITY, (
+        "cannot perform reduction function max on tensor with "
+        "no elements because the operation does not have an identity"));
+      TORCH_CHECK(ord != -INFINITY, (
+        "cannot perform reduction function min on tensor with "
+        "no elements because the operation does not have an identity"));
+    }
+  }
+
+  ScalarType out_dtype = opt_dtype.has_value() ? opt_dtype.value() : toValueType(self.scalar_type());
+  Tensor result_;
+  Tensor self_;
+
+  if (self.device().type() == c10::kCPU && std::abs(ord) == INFINITY) {
+    // TODO: For some reason, for complex CPU, at::abs() gives a different
+    // result than the absolute operation within linalg_vector_norm_stub. This
+    // causes the backward method to fail. Current workaround is to use
+    // at::abs() directly in this case.  I'm not sure what the issue, but it
+    // should be fixed before PR can be merged.
+    self_ = self.to(in_dtype).abs();
+    in_dtype = toValueType(in_dtype);
+  } else {
+    self_ = self;
+  }
+  auto iter = make_reduction("vector_norm", result_, self_, dim, keepdim, in_dtype, out_dtype);
+
+  linalg_vector_norm_stub(iter.device_type(), iter, ord);
+
+  if (result_.dtype() != out_dtype) {
```

Comment:
I'm confused about what's going on here -- what if result is undefined?

In particular, what's the problem with just passing result directly to make_reduction? Is it that reductions don't support complex -> real?

Reply:
You're right, there is a simpler way to do this. `result_.dtype() != out_dtype` is always false anyway, so this was a mistake.

I have a better implementation on my local branch, but I haven't pushed it yet because I've been benchmarking this function and speeding things up to at least match what we get with `torch.norm`. Once that's done, I'll push my updates and make a post about performance